errors/errors_test.go

```diff
@@ -810,3 +813,77 @@ func TestErrorBuilder_Build(t *testing.T) {
 		})
 	})
 }
+
+// This test ensures that all the error details google provides are covered in our switch case
+// in errors.go. If google adds an error detail, this test should fail, and we should add
+// that specific error detail to the switch case
+func TestEnsureAllErrDetailsCovered(t *testing.T) {
+	packagePath := "google.golang.org/genproto/googleapis/rpc/errdetails"
+
+	// Load the package
+	cfg := &packages.Config{Mode: packages.NeedTypes | packages.NeedTypesInfo}
+	pkgs, err := packages.Load(cfg, packagePath)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if packages.PrintErrors(pkgs) > 0 {
+		t.Errorf("ensure package is correct: %v", packages.ListError)
+	}
+
+	// This is hard-coded from the switch statement in errors.go to ensure we stay up
+	// to date on the error types we support, and to ensure we update our supported
+	// error types when google adds to their error details
+	mySwitchTypes := []string{
```

Comment:
Is there a way this can not be hardcoded? Maybe using reflection or the ast package?

Another option could be not to hardcode this list. Assuming we don't care if we have extra cases in the `switch`:

1. Export the `switch` statement in its own function so it's easier to test
2. Iterate on all the types exported by the Google's package
3. Invoke the function
4. Confirm the function can handle that case

Reply:
I now dynamically grab the values using the ast pkg. Its very ugly. I've not used the pkg before, so if you can make it cleaner I'm open to changing it, but it gets the job done dynamically and I added comments on the ast bits since its not pretty.

Reply:
This looks pretty nice to me now @cicoyle . I've only used the ast pkg once so I'm no expert myself, but if it works... :) Thanks for creating this test, looks really good work